* Fractal: The fractal layout is designed to display only the insertion vertices in an alignment. Vertices with a single inserted nucleotide (A, C, G, or T) are positioned in the center of the corresponding quadrant shown in Figure 2. Specifically, A is in the top-left quadrant, C is in the top-right quadrant, G is in the bottom-left quadrant, and T is in the bottom-right quadrant. For vertices with two inserted nucleotides, they are positioned in the center of the corresponding square shown in Figure 3. The pattern for two-nucleotide insertions starting with A involves subdividing the A square into four quadrants and arranging AA, AC, AG, and AT based on their second nucleotide in the same order as the original A, C, G, T squares. Specifically, AA is in the top-left quadrant, AC is in the top-right quadrant, AG is in the bottom-left quadrant, and AT is in the bottom-right quadrant. This pattern can be extended to arbitrarily large kmers. For a kmer $s_1 s_2 \cdots s_k$, the corresponding square will be obtained by subdividing the square for $s_1 s_2 \cdots s_{k-1}$ into four quadrants and choosing the quadrant in the same way as before based on $s_k$. This layout is similar to the insertion side of the universal layout, but the inserted sequences are arranged in a two-dimensional form instead of linearly from left to right. The final coordinates of the vertices are always in the range [-1, 1], which is relevant when setting the `--plot_range_x` and `--plot_range_y` parameters.

![Fractal layout 1-length insertion placement](figures/fractal_1.png)
*Figure 2: Fractal layout vertex placement for insertions of length 1. The vertices is placed in the certer of the square corresponding to its inserted string.*

![Fractal layout 2-length insertion placement](figures/fractal_2.png)
*Figure 2: Fractal layout vertex placement for insertions of length 2. Each square for the length 1 insertions are subdivided into 4 new squares. The vertices is placed in the certer of the square corresponding to its inserted string.*
